path_oidc.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	roleName := d.Get("role").(string)
+	if roleName == "" {
+		roleName = config.DefaultRole
+		if roleName == "" {
+			return logical.ErrorResponse("missing role"), nil
+		}
+	}
+
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return logical.ErrorResponse("missing redirect_uri"), nil
+	}
+
+	if role != nil && role.RoleType == "oidc" {
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		if len(role.AllowedRedirectURIs) == 0 ||
+			strutil.StrListContains(role.AllowedRedirectURIs, redirectURI) {
+
+			// "openid" is a required scope for OpenID Connect flows
+			scopes := append([]string{oidc.ScopeOpenID}, role.OIDCScopes...)
+
+			// Configure an OpenID Connect aware OAuth2 client
+			oauth2Config := oauth2.Config{
+				ClientID:     config.OIDCClientID,
+				ClientSecret: config.OIDCClientSecret,
+				RedirectURL:  redirectURI,
+				Endpoint:     provider.Endpoint(),
+				Scopes:       scopes,
+			}
+
+			stateID, nonce, err := b.createState(roleName, redirectURI)
+			if err != nil {
+				return nil, errwrap.Wrapf("error generating OAuth state: {{err}}", err)
+			}
+
+			authCodeURL = oauth2Config.AuthCodeURL(stateID, oidc.Nonce(nonce))
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"auth_url": authCodeURL,
+		},
+	}
+
+	return resp, nil
+}
+
+// createState make an expiring state object, associated with a random state ID
+// that is passed throughout the OAuth process. A nonce is also included in the
+// auth process, and for simplicity will be identical in length/format as the state ID.
+func (b *jwtAuthBackend) createState(rolename, redirectURI string) (string, string, error) {
+	randstr, err := base62.Random(2 * stateLength)
+	if err != nil {
+		return "", "", err
+	}
+
+	stateID, nonce := randstr[0:stateLength], randstr[stateLength:]
+
+	b.l.Lock()
+	b.oidcStates[stateID] = &oidcState{
+		rolename:    rolename,
+		expiration:  time.Now().Add(oidcStateTimeout),
+		nonce:       nonce,
+		redirectURI: redirectURI,
+	}
+	b.l.Unlock()
+
+	return stateID, nonce, nil
+}
+
+// verifyState tests whether the provided state ID is valid and returns the
+// associated state object if so. A nil state is returned if the ID is not found
+// or expired. The state should only ever be retrieved once and is deleted as
+// part of this request.
+func (b *jwtAuthBackend) verifyState(state string) *oidcState {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	s := b.oidcStates[state]
+	if s != nil && time.Now().After(s.expiration) {
+		s = nil
+	}
+
+	delete(b.oidcStates, state)
+
+	return s
+}
+
+// stateGC will start a goroutine which periodically deletes all expired states
+// that were never removed via the normal request process.
+func (b *jwtAuthBackend) stateGC() {
+	go func() {
```

Comment:
Is there any way to tie this func to something that will cleanly shut it down if the `jwtAuthBackend` is terminated? Like what if it were unmounted, or a request context were cancelled? Ideally there'd be a channel of some sort being passed in.